server/container_create_linux.go

```diff
@@ -411,6 +414,30 @@ func (s *Server) mountImage(ctx context.Context, specgen *generate.Generator, im
 
 	log.Infof(ctx, "Image mounted to: %s", mountPoint)
 
+	// Kubernetes already verifies that ImageSubPath:
+	// - is a relative path
+	// - does not contain any '..'
+	//
+	// We cannot create non-existing paths within the image volume because
+	// they're mounted read-only.
+	mountPoint, err = securejoin.SecureJoin(mountPoint, m.ImageSubPath)
+	if err != nil {
+		return nil, fmt.Errorf("secure join final mount path: %w", err)
+	}
+
+	var st unix.Stat_t
+	if err := unix.Stat(mountPoint, &st); err != nil {
+		// This is a user-facing error message from a kubelet event,
+		// means we should make it as meaningful as possible.
+		if os.IsNotExist(err) {
+			return nil, fmt.Errorf("sub path %q does not exist in image volume %q", m.ImageSubPath, m.Image.Image)
```

Comment:
this is a potential security issue.

There is a TOCTOU here, since the source can change until the OCI runtime uses it.

e.g. if the same volume is used by two different containers, a container already having access to the volume can replace the directory with a symlink.

I ended up doing this in Podman: https://github.com/containers/podman/pull/17992

Reply:
Hm, I guess we could also use `securejoin.OpenInRoot` for that: https://github.com/cyphar/filepath-securejoin/blob/f277210982c593f9015a66100c67b20c957c24fe/open_linux.go#L43-L50